20231214-listUsers-api.md

```diff
@@ -0,0 +1,887 @@
+
+### Expected Outcome
+
+Two new core APIs are added to the [OpenFGA API](https://github.com/openfga/api) and implementations of these new APIs in the [OpenFGA server](https://github.com/openfga/openfga).
+
+## What it is
+[what-it-is]: #what-it-is
+
+Given an `object`, `relation`, and one or more user/subject provided filters, return the concrete/terminal set of user objects matching at least one of the user filters that have that relationship with the object.
+
+More formally, given the following model and relationship tuples:
+
+```
+model
+  schema 1.1
+
+type user
+type cat
+
+type group
+  relations
+    define member: [user, group#member]
+
+type document
+  relations
+    define viewer: [cat, user, group#member]
+```
+
+| object     | relation | user             | condition_name | context |
+|------------|----------|------------------|----------------|---------|
+| document:1 | viewer   | user:anne        | null           | null    |
+| document:1 | viewer   | group:eng#member | null           | null    |
+| group:eng  | member   | group:fga#member | null           | null    |
+| group:fga  | member   | user:jon         | null           | null    |
+
+ListUsers and StreamedListUsers would behave as follows:
+
+> ℹ️ For the sake of brevity the examples below focus on the ListUsers unary RPC variant. The difference with the StreamedListUsers API is the streaming semantics of the response. Instead of returning an array, the streaming variant returns each result one at a time as they are "discovered".
+
+**Example 1**
+`user` object types are both directly and indirectly related to `document#viewer` relationships. The indirect relationships come through the direct relationships involving `group#member` (e.g. users can directly view a document or if the user is part of a group that can view the document)
+```
+ListUsers({
+    object: "document:1",
+    relation: "viewer",
+    user_filters: [{type: "user"}]
+}) --> 
+{
+  users: [],
+  wildcard_types: ["user:anne", "user:jon"]
+}
+```
+
+**Example 2**
+Looking at the type restrictions defined in the model, `group` objects are not directly related whatsoever to `document#viewer` relationships. So this example returns no results.
+```
+ListUsers({
+    object: "document:1",
+    relation: "viewer",
+    user_filter: [{type: "group"}]
+}) --> {
+  users: [],
+  wildcard_types: []
+}
+```
+
+**Example 3**
+You can see in the model that `group#member` relationships are directly related to `document#viewer`. So this example returns the `group:eng#member` direct relationship and the `group:fga#member` relationship indirectly through `group:eng#member`.
+```
+ListUsers({
+    object: "document:1",
+    relation: "viewer",
+    user_filter: [
+      {
+        type: "group",
+        relation: "member"
+      }
+    ]
+}) --> 
+{
+  users: ["group:eng", "group:fga"]
+  wildcard_types: []
```

Comment:
do we need this? Why not return this in the first array

Reply:
IMO it makes it easier for parsing and it makes it more obvious that you need to look at the wildcard. Imagine a response with 1000 results AND a wildard thrown in the middle. It will be harder for me to debug 🤔

Reply:
But this will change once we start allowing multiple filters, as the responses might be drastically different and you have to parse them anyway.

E.g. when you are getting back
```
users: ["user:anne", "group:fga#member", "employee:*"].
```
or, as I would prefer it
```
users: [{
  type: user,
  id: anne
}, {
  type: group,
  id: fga,
  relation: member
}, {
  type: employee
  wildcard: {}
}]
```

Some may be wildcards, some may be usersets, some may be objects - each with a different semantic meaning they would need to parse

Reply:
The wildcard is a special case because it does not tie to an actual user. For both the highlighted use cases (share dialog, notifications), if a wildcard was returned, the user would necessarily need to iterate and filter over the set. And that's _if_ the developer is aware that wildcards are returned in such a way. 

The proposed response shape you provided ensures that all developers will need to iterate over the results and parse based on specific criteria. The ergonomics of that are poor when compared to separating the response cases.